### ISSUE TITLE
HTTP/2: REFUSED_STREAM causes delay in sending frames

### ISSUE THREAD

**Issue by JamesNK** [dotnet/runtime#41369]
This is a follow up to https://github.com/dotnet/runtime/issues/40249. That issue was mostly resolved by limiting SocketsHttpHandler to sending 100 concurrent requests until the HTTP/2 connection is fully negotiated (i.e. the client has received a settings frame from the server).

However there still appears to be an underlying bug if the following conditions are met:
1. The server is configured with a SETTINGS_MAX_CONCURRENT_STREAMS that is lower than 100, i.e. 50.
2. SocketsHttpHandler sends more than SETTINGS_MAX_CONCURRENT_STREAMS number of requests (i.e. 250) before HTTP/2 connection is fully negotiated.
3. The 250 requests are long running.

In this situation SocketsHttpHandler appears to send frames to the server slowly. The server logs show there is exactly a 1 second delay between each burst of activity. Eventually there is a 5 second pause between the server receiving a HEADER frame (which it refuses) and it receiving a DATA frame. Kestrel will forget about a refused stream after 5 seconds. Getting a DATA frame with an unknown stream ID is a connection violation so the server closes the connection.

Repo: [RefusedStreamRepo.zip](https://github.com/dotnet/runtime/files/5127021/RefusedStreamRepo.zip)

Broken:

1. dotnet run -p WebApplication61
2. dotnet run -p ConsoleApp59
3. The console app sends 250 requests immediately. The server console logs activity, including the frames it receives. You see an initial burst of activity of incoming frames. After that initial burst, frames are received from the client once per second.
4. After a couple of minutes of slowly receiving frames the server receives a DATA frame on an unknown stream. That stream has been forgotten about because it was refused more than 5 seconds again. Kestrel kills the connection.

Working:

1. Open `WebApplication61\Program.cs` and change MaxStreamsPerConnection from 50 to 100
2. dotnet run -p WebApplication61
3. dotnet run -p ConsoleApp59
4. The console app sends 250 requests immediately. In less than a second the server has received all 250 requests, and they all complete successfully.

Having all 3 conditions to create this problem should be rare. By default servers have a SETTINGS_MAX_CONCURRENT_STREAMS value over 100 or greater, but the client behavior when this issue does happen is pretty bad.

**Comment by scalablecory** [comment]
Based on the description it's not clear if this is an issue with HttpClient or Kestrel, or both, but we should figure this out for 6.0. It would also be helpful to know if HttpClient exhibits this behavior against any other HTTP/2 servers.

**Comment by BrennanConroy** [comment]
I'm investigating a Kestrel test issue and am hitting this. The client side behavior isn't great in this case:
```
System.Net.Http.HttpRequestException : An error occurred while sending the request.
    ---- System.IO.IOException : The request was aborted.
    -------- System.ObjectDisposedException : Cannot access a disposed object.
    Object name: 'System.Net.Sockets.NetworkStream'.
```

I haven't tried against other servers, but it looks like according to the spec, endpoints have a little bit of flexibility in how they handle frames after closing the stream before they send STREAM_CLOSED and close the connection. In Kestrel's case it will wait 5 seconds before forgetting about a stream:
```
If a DATA frame is received whose stream is not in "open" or "half-closed (local)" state, the recipient MUST respond with a stream error (Section 5.4.2) of type STREAM_CLOSED.
```
and
```
closed:
...
If this state is reached as a result of sending a RST_STREAM frame, the peer that receives the RST_STREAM might have already sent — or enqueued for sending — frames on the stream that cannot be withdrawn. An endpoint MUST ignore frames that it receives on closed streams after it has sent a RST_STREAM frame. An endpoint MAY choose to limit the period over which it ignores frames and treat frames that arrive after this time as being in error.
```

I have also turned on connection logging in Kestrel to see which side is to blame for the large gap in the RST_STREAM frame being sent and the DATA frame being received. My preliminary results show that Kestrel read the bytes, but didn't process them for over 5 seconds. I can't explain the 1 second batches yet, but those don't seem to be caused by the client. From the logs the last read happened at 0.687s and the Data frame that causes the connection closure is at 6.262s. In between all that, there are the strange batches of processing and sending various frames to the client.

TLDR:
It seems like there are 2 issues here.
1. The client should close more gracefully, an ODE isn't a great error to surface to the user
2. Kestrel has some odd processing behavior that needs to be investigated

**Comment by BrennanConroy** [comment]
>     2\. Kestrel has some odd processing behavior that needs to be investigated

Good news, I figured out the Kestrel issue and am working on a PR for it.

This issue should probably be re-titled or replaced by another issue for the client-side exception.
